src/event-log/event-log-level.ts

```diff
@@ -15,101 +18,114 @@ type EventLogLevelConfig = {
   createLevelDatabase?: typeof createLevelDatabase,
 };
 
-const WATERMARKS_SUBLEVEL_NAME = 'watermarks';
 const CIDS_SUBLEVEL_NAME = 'cids';
 
 export class EventLogLevel implements EventLog {
-  config: EventLogLevelConfig;
-  db: LevelWrapper<string>;
   ulidFactory: ULIDFactory;
+  index: IndexLevel<Event>;
 
   constructor(config?: EventLogLevelConfig) {
-    this.config = {
+    this.index = new IndexLevel({
       location: 'EVENTLOG',
       createLevelDatabase,
       ...config,
-    };
-
-    this.db = new LevelWrapper<string>({
-      location            : this.config.location!,
-      createLevelDatabase : this.config.createLevelDatabase,
-      valueEncoding       : 'utf8',
     });
+
     this.ulidFactory = monotonicFactory();
   }
 
   async open(): Promise<void> {
-    return this.db.open();
+    return this.index.open();
   }
 
   async close(): Promise<void> {
-    return this.db.close();
+    return this.index.close();
   }
 
   async clear(): Promise<void> {
-    return this.db.clear();
+    return this.index.clear();
   }
 
-  async append(tenant: string, messageCid: string): Promise<string> {
-    const tenantEventLog = await this.db.partition(tenant);
-    const watermarkLog = await tenantEventLog.partition(WATERMARKS_SUBLEVEL_NAME);
-    const cidLog = await tenantEventLog.partition(CIDS_SUBLEVEL_NAME);
-
+  /**
+   * Appends Events to the EventLog in the order they are appended using a ulid watermark.
+   * The indexes are used to query the event-log for specific events.
+   *
+   * @param tenant
+   * @param messageCid the messageCid of the message that is being appended to the log.
+   * @param indexes property and value indexes to use for querying potential events.
+   * @returns the ulid watermark generated during this operation.
+   */
+  async append(tenant: string, messageCid: string, indexes: { [key:string]: unknown }): Promise<string> {
     const watermark = this.ulidFactory();
-
-    await watermarkLog.put(watermark, messageCid);
-    await cidLog.put(messageCid, watermark);
-
+    // A reverse lookup in order to delete by messageCid
+    // We are using the watermark as the key for this index.
+    // When deleting by messageCid we need to look up the watermark.
+    const cidLogIndex = await this.messageCidPartition(tenant);
+    await cidLogIndex.put(messageCid, watermark);
+
+    const event:Event = { messageCid, watermark };
+    await this.index.put(tenant, watermark, event, { ...indexes, watermark });
```

Comment:
Design: I am not 100% convinced that the motivation behind changing the `index.put(...)` method to incorporate "value" as input is really needed (though I have yet to go through the entire PR), especially given that we already have a CID -> watermark table, so seems like the same pattern used by the message store can be followed.

We should probably discuss this one. To be clear, the potential change here isn't much, a simplification even.

Reply:
This is done and simplified!